hypothesis-python/src/_hypothesis_globals.py

```diff
@@ -0,0 +1,28 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Copyright the Hypothesis Authors.
+# Individual contributors are listed in AUTHORS.rst and the git log.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""
+Module for globals shared between plugin(s) and the main hypothesis module, without
+depending on either. This file should have no imports outside of stdlib.
+"""
+
+import os
+
+in_initialization = 1
+"""If nonzero, indicates that hypothesis is still initializing (importing or loading
+the test environment). `import hypothesis` will cause this number to be decremented,
+and the pytest plugin increments at load time, then decrements it just before the test
+session starts. However, this leads to a hole in coverage if another pytest plugin
+imports hypothesis before our plugin is loaded. HYPOTHESIS_EXTEND_INITIALIZATION may
+be set to pre-increment the value on behalf of _hypothesis_pytestplugin, plugging the
+hole."""
```

Comment:
Incrementing and decrementing a counter has my threadsafety senses tingling.  OK, we formally claim that Hypothesis is _not_ threadsafe, but I think parallel test runners are common enough to suggest using a threadlocal variable for this.

Reply:
I assume imports are always serialized. I admit I haven't verified this, but anything else would be a can of worms.

So any races would have to be in the pytest plugin, and even with a threadlocal variable we'd need to know the threaded hook semantics to make it right. I couldn't find any example of a threaded runner to check, the only one is pytest-parallel which is abandoned.

How about this: We loosen the balancing - only increment the counter at import time (safe) but decrement it for every session start. That would make all threads see at least one decrement, even if the exact number is racy. Then check for `is_initialized <= 0` instead of `== 0`. In other words, it would be racy but it would fail safe by not showing warnings. What do you think?

Reply:
Looking through it more carefully, I think you're right and it's not actually worth doing anything here - the GIL protects us from all the really nasty possibilities, and the worst case we'd observe is some missed-or-extra warnings before initialization finished.  My bad 😅